src/Data/AIG/CompactGraph.hs

```diff
@@ -0,0 +1,444 @@
+       maxVar  <- newIORef maxv
+       inputs  <- newIORef inps
+       andMap  <- newIORef gates
+       return (Network CompactGraph {..} outs)
+
+  trueLit  _g = CompactLit 1
+  falseLit _g = CompactLit 0
+
+  newInput g =
+    do v <- newVar g
+       modifyIORef' (inputs g) (v :)
+       return (varToLit v)
+
+  and g x y =
+    do v <- newVar g
+       let l = max x y
+           r = min x y
+       modifyIORef' (andMap g) $ Map.insert v (l, r)
+       return (varToLit v)
+
+  inputCount g = length <$> readIORef (inputs g)
+
+  -- | Get input at given index in the graph.
+  getInput _g _i =
+    fail "Function getInput not implemented for CompactGraph"
+
+  writeAiger fp ntk = writeAigerWithLatches fp ntk 0
+
+  writeAigerWithLatches fp (Network g outs) latches =
+    withFile fp WriteMode $ \h ->
+    do var <- readIORef (maxVar g)
+       ins <- reverse <$> readIORef (inputs g)
+       gateMap <- readIORef (andMap g)
+       let vm = mkVarMap ins gateMap
+           format = Binary
+       writeHeader h format var ins latches outs gateMap
+       writeInputs h format latches vm ins
+       writeLatches h format latches vm ins outs
+       writeOutputs h latches vm outs
+       writeAnds h format vm gateMap
+
+  writeCNF g out fp =
+    withFile fp WriteMode $ \h ->
+    do Var var <- readIORef (maxVar g)
+       ins <- reverse <$> readIORef (inputs g)
+       gateMap <- readIORef (andMap g)
+       let vm = mkVarMap ins gateMap
+           nvars = fromIntegral var + 1
+           nclauses = (3 * Map.size gateMap) + 2
+           litToCNF lit =
+             case Map.lookup (litToVar lit) vm of
+               Just (Var v) ->
+                 do let n = fromIntegral v + 1
+                    return $ if litNegated lit then (-n) else n
+               Nothing -> fail $ "Literal not found: " ++ show lit
+           putClause lits =
+             hPutBBLn h $ (bsUnwords . map BS.intDec) lits <> " 0"
+       hPutBBLn h $ bsUnwords [ "p", "cnf"
+                              , BS.intDec nvars
+                              , BS.intDec nclauses
+                              ]
+       forM_ (Map.assocs gateMap) $ \(v, (ll, rl)) ->
+         do n <- litToCNF (varToLit v)
+            li <- litToCNF ll
+            ri <- litToCNF rl
+            -- 3 clauses for each gate
+            putClause [-n, li]
+            putClause [-n, ri]
+            putClause [n, -li, -ri]
+       ovar <- litToCNF out
+       -- 2 more clauses
+       putClause [ovar]
+       putClause [-1]
+       return [2 .. length ins + 1]
+
+  checkSat _g _l =
+    -- Could call out to ABC for this?
+    fail "Cannot SAT check graphs in the CompactGraph implementation"
+
+  cec _g1 _g2 =
+    -- Could call out to ABC for this?
+    fail "Cannot CEC graphs in the CompactGraph implementation"
+
+  -- | Evaluate the network on a set of concrete inputs.
+  evaluator _g _xs =
+    fail "evaluator not implemented"
```

Comment:
I think you can implement a naive version of this using `abstractEvaluateAIG`.  It won't be especially fast, but it should work.

Reply:
Yeah, like `getInput` it's not currently used but is probably still worth including. I'll do it.